### PR TITLE
Correct error in function Modelica.Media.Air.ReferenceMoistAir.Utilities.pd_pTX_der

### LINKED ISSUE #3666: ReferenceMoistAir, substances mass conservation issue
Hello,

I would like to report an issue with the Modelica.Media.Air.ReferenceMoistAir media.

From unexpected simulation results, I implemented the following simple model (cf. enclosed model): a constant mass flow source (Modelica.Fluid.Sources.MassFlowSource_T) with X = {0, 1} connected to a constant volume tank (Modelica.Fluid.Vessels.ClosedVolume) with X = {0.5, 0.5} @ t = 0. The tank water mass is expected to remain the same and the total mass was expected to increase linearly. Instead, the tank water mass increased and the total fluid mass increasing beyond expectation and beyond the water mass increase. 

[XBalance.zip](https://github.com/modelica/ModelicaStandardLibrary/files/5553752/XBalance.zip)

The issue has been observed with Dymola 2020 using Modelica Standard Library - Version 3.2.3 and confirmed using Modelica Standard Library - Version 4.0.0. As a comparison, using Modelica.Media.IdealGases.MixtureGases.AirSteam, the observed error is more likely due to computing small error propagation.

Quick investigations pointed to ReferenceMoistAir density and derivated functions.

Please let me know if problem if the problem has already been reported and if solution has been developed.

Thank you !

### PR DESCRIPTION
**Pull request by beutlich**
Resolves #3666 as proposed by @GaelEnee.

### CODE REVIEW

**Review by hubertus65 — CHANGES_REQUESTED** [review]
See my latest comment in the conversation of this PR: I think the derivative is wrong, the correct one is proposed in the original thread.

**Comment by TManikantan** [comment]
@hubertus65 @thorade  a gentle reminder this issue remains unresolved, Restarting the process of solving it would be really needed.

**Comment by TManikantan** [comment]
@hubertus65 would you please look into this ticket?

**Comment by HansOlsson** [comment]
Why do we write such functions at all? Modelica has smoothOrder that automatically derive such a derivative-function from the original, and it would automatically ensure that any other differences disappear - e.g., handling of xws==-1.

Thus my proposal is to just remove the functions pd_pTX_der, h_pTX_der, rho_pTX_der, and u_pTX_der and replace them with smoothOrder=1.

That would avoid such errors, and reduce the maintenance burden.

**Comment by beutlich** [comment]
By applying smoothOrder=1 and avoiding the explicit derivative functions (see https://github.com/modelica/ModelicaStandardLibrary/commit/e3c45f202573f81f6f986f4321a42fbd1b77e8ef) I get a differentiation error in Dymola 2023

```
Translation of pd_der_test_smoothOder:

The DAE has 8 scalar unknowns and 8 scalar equations.

Failed to differentiate the equation
var_from_function = Modelica.Media.Air.ReferenceMoistAir.Utilities.pd_pTX(p, T, X);

in order to reduce the DAE index.

Cannot find differentiation function:
Modelica.Media.Air.ReferenceMoistAir.Utilities.pd_pTX(p, T, X)
with respect to time

Failed to reduce the DAE index.

Translation aborted.

ERRORS have been issued.
```

for the test model:

```mo
model pd_der_test_smoothOrder
  replaceable package Medium = Modelica.Media.Air.ReferenceMoistAir;

  Real der_T = 300;
  Real T = 73.15 + der_T * time;
  Real der_p = -1e5;
  Real p = 1e7 + der_p * time;
  Real[Medium.nS] X = Medium.reference_X;

  Real var_from_function = Medium.Utilities.pd_pTX(p, T, X);
  Real timederiv = der(var_from_function);

  annotation (experiment(StopTime=100));
end pd_der_test_smoothOrder;
```

Is there anything I am missing?

**Comment by beutlich** [comment]
> > Is there anything I am missing?
> 
> Possibly, I'll have to make a branch and test that.

You can pull my branch: https://github.com/beutlich/ModelicaStandardLibrary/tree/issue3666_avoid-derivative-functions

**Comment by TManikantan** [comment]
@HansOlsson @beutlich did you get around the solution. Kindly commit the fix if you have arrived at the solution.
Many Thanks

**Comment by HansOlsson** [comment]
> > > Is there anything I am missing?
> > 
> > 
> > Possibly, I'll have to make a branch and test that.
> 
> You can pull my branch: https://github.com/beutlich/ModelicaStandardLibrary/tree/issue3666_avoid-derivative-functions

Hm...
TL;DR: Seems we have to delay smoothOrder-change, and do it as another PR.

More detailed:

Dymola currently cannot differentiate this for two reasons:
* It cannot find derivative of Tlim in `Modelica.Media.Air.ReferenceMoistAir.Utilities.pds_pT` - note that this derivative isn't needed. It can be worked around by replacing` T<=Tlim` by a `Boolean notAboveTlim;` and instead of computing `Tlim := ...;` compute `notAboveTlim:= T<=...;`
* No derivative for `Modelica.Media.Air.ReferenceMoistAir.Utilities.f_pT(p, T)` (due to limitations for handling functions as arguments to functions)- as far as I can see the current implementation just assumes this is 1 in the derivative-function! I don't know how correct that is.

I don't know what other tools can do, I could imagine similar limitations - but I think the limitations can be overcome (for functions as argument to functions it depends on the functions). The problematic part for MSL is that the 2nd change is another bug (or simplification) so a correction needs more testing; which is another reason this should be a separate PR.

**Comment by TManikantan** [comment]
> > > > Is there anything I am missing?
> > > 
> > > 
> > > Possibly, I'll have to make a branch and test that.
> > 
> > 
> > You can pull my branch: https://github.com/beutlich/ModelicaStandardLibrary/tree/issue3666_avoid-derivative-functions
> 
> Hm... TL;DR: Seems we have to delay smoothOrder-change, and do it as another PR.
> 
> More detailed:
> 
> Dymola currently cannot differentiate this for two reasons:
> 
> * It cannot find derivative of Tlim in `Modelica.Media.Air.ReferenceMoistAir.Utilities.pds_pT` - note that this derivative isn't needed. It can be worked around by replacing` T<=Tlim` by a `Boolean notAboveTlim;` and instead of computing `Tlim := ...;` compute `notAboveTlim:= T<=...;`
> * No derivative for `Modelica.Media.Air.ReferenceMoistAir.Utilities.f_pT(p, T)` (due to limitations for handling functions as arguments to functions)- as far as I can see the current MSL-implementation just assumes this is 1 in the derivative-function! I don't know how correct that is.
> 
> I don't know what other tools can do, I could imagine similar limitations - but I think the limitations can be overcome (for functions as argument to functions it depends on the functions). The problematic part for MSL is that the 2nd change is another bug (or simplification) so a correction needs more testing; which is another reason this should be a separate PR.

if i am correct ,what i could understand by reading the comment is that you propose to go ahead with the changes suggested by@hubertus65 and then do the simplification work in a different PR. Kindly correct me if i am wrong. Many thanks..

**Comment by HansOlsson** [comment]
> if i am correct ,what i could understand by reading the comment is that you propose to go ahead with the changes suggested by@hubertus65 and then do the simplification work in a different PR. Kindly correct me if i am wrong. Many thanks..

Yes.

**Comment by TManikantan** [comment]
@hubertus65 @thorade Kindly look into the changes done to the fix as per your comments.

**Review by thorade — APPROVED** [review]

**Comment by thorade** [comment]
I manually derived the expression and arrived at the same result. I also tested with the test model, and the error becomes much smaller, but not zero.

**Review by hubertus65 — APPROVED** [review]
